pkg/querier/queryrange/results_cache.go

```diff
@@ -41,10 +41,10 @@ type resultsCache struct {
 }
 
 // NewResultsCacheMiddleware creates results cache middleware from config.
-func NewResultsCacheMiddleware(logger log.Logger, cfg ResultsCacheConfig, limits Limits) (Middleware, error) {
+func NewResultsCacheMiddleware(logger log.Logger, cfg ResultsCacheConfig, limits Limits) (Middleware, cache.Cache, error) {
```

Comment:
I think I would add a `Stop()` method to the `Middleware` return rather than exposing the implementation.

Reply:
> I think I would add a `Stop()` method to the `Middleware` return rather than exposing the implementation.

I agree, let me take another swing at it.

Reply:
@bboreham, I looked at this once again. `Frontend` does not have direct access to `queryrange.Middleware` member. Instead, it has `roundTripper` interface member of type `http.RoundTripper`. `queryrange.NewRoundTripper` returns `queryrange.roundTripper` object that implements `http.RoundTripper` but hides `Middleware`s internally. So, adding a new method to `queryrange.Handler` would break parity between `http.RoundTripper` and `queryrange.roundTripper`, hence the design.
I would argue that Cache is a special component that requires separate treatment, different from `RoundTripper`.
I'm open to an idea of adding `Close' method to `queryrange.Middleware` but IMO this PR in its current form offers sufficient and isolated changes.

Reply:
I had a read through this and agree with @dmitsh 's findings. It might be a good point to revaluate some of the logic  in order to refactor away the implementation, but that seems like a larger effort than just addressing the issue at hand.